# Darwindow

Darwindow is a small tool, consisting of a Unix and a R script, to calculate and plot population-genetic estimates on a sliding-window basis, using as input a gVCF file  (i.e., a vcf-file containing information on both monomorphic and polymorphic sites for the entire genome). 

Darwindow is particularly useful to perform run of homozygosity (ROH) analyses. The advantage of Darwindow relative to other ROH-detection software is that Darwindow allows the user to evaluate the accuracy of the ROH calls through visual examination of heterozygosity levels in regions marked as ROH (grey rectangles in plots with prefix 'He_withROH_linechart', generated by the runindscaffold function).

The average genome-wide heterozygosity estimates produced by Darwindow are near-identical to the estimates obtained with 'bcftools stats -s -', assuming that indels have been removed from the input vcf-file using the command 'zgrep -v "INDEL" data.vcf.gz | gzip > data.noindels.vcf.gz' (this step is included in Darwindow pipeline).

In order to run Darwindow, follow the instructions in the top section of the VCF_darwindow.sh script.
This will generate intermediate file(s) containing window-based counts (such as number of homozygous and number of heterozygous sites per individual per window).
These window-based counts can subsequently be imported and analysed in R, using the script 'VCF_darwindow.plotinR.txt'.
The top section of this R script contain an example run with commands.






